bgbb/sql/sql_utils.py

```diff
@@ -0,0 +1,219 @@
+    # mod_ho_ev_str
+    (r.model_start_date_str, r.ho_start_date_str, r.ho_last_date_str
+     ) = map(to_s3_fmt, mod_ho_ev)
+
+    test_ho_range(r.ho_start_date, r.ho_last_date, HO_WIN)
+    test_model_range(model_start_date=r.model_start_date,
+                     ho_start_date=r.ho_start_date,
+                     MODEL_WIN=MODEL_WIN)
+
+    # r.__dict__.update(locals())
+    return r
+
+
+base_query = '''
+with cid_day as (
+    SELECT
+        C.client_id
+        , C.submission_date_s3
+        , from_unixtime(unix_timestamp(C.submission_date_s3, 'yyyyMMdd'),
+                        'yyyy-MM-dd') AS sub_date
+    FROM clients_daily C
+    WHERE
+        app_name = 'Firefox'
+        AND channel = 'release'
+      {sample_comment}  AND sample_id in ({sample_ids})
+)
+
+-- clients_daily aggregates from window *before*
+-- the holdout date
+, cid_model as (
+    SELECT
+        C.client_id
+        , MIN(C.sub_date) Min_day
+        , MAX(C.sub_date) AS Max_day
+        , COUNT(*) X
+    FROM cid_day C
+    WHERE
+      C.submission_date_s3 >= '{model_start_date_str}'
+      AND C.submission_date_s3 < '{ho_start_date_str}'
+    GROUP BY 1
+)
+
+, cid_holdout as (
+    SELECT
+        C.client_id
+        , COUNT(*) N_holdout
+    FROM cid_day C
+    WHERE
+      C.submission_date_s3 >= '{ho_start_date_str}'
+      AND C.submission_date_s3 <= '{ho_last_date_str}'
+    GROUP BY 1
+)
+
+, rec_freq as (
+    SELECT
+        C.client_id
+        , datediff(C.Max_day, Min_day) Recency
+        , X - 1 Frequency
+        , datediff('{ho_start_date}', Min_day) - 1 T
+        , C.Max_day
+        , C.Min_day
+    FROM cid_model C
+)
+
+, rec_freq_holdout as (
+  SELECT R.*
+        , coalesce(H.N_holdout, 0) N_holdout
+  FROM rec_freq R
+  LEFT JOIN cid_holdout H
+    ON R.client_id = H.client_id
+)
+
+SELECT * FROM {qname}
+'''
+
+
+def mk_rec_freq_q(q=None, holdout=False, model_start_date_str=None, pcd=None,
+                  sample_ids="'1'", **k):
+    """
+    holdout: pull # of returns in holdout period?
+    """
+    qname = "rec_freq_holdout" if holdout else "rec_freq"
+    kw = dict(
+        model_start_date_str=model_start_date_str,
+        sample_ids=sample_ids,
+        sample_comment='' if sample_ids else '--',
+        qname=qname,
+        pcd=pcd
+    )
+    kw.update(k)
+    kw = {k: v for k, v in kw.items() if v is not None}
+    return q.format(**kw)
+
+
+def run_rec_freq_spk(rfn_base_query=base_query, HO_WIN=14, MODEL_WIN=90,
+                     holdout=False, sample_ids="'1'",
+                     ho_start='2018-08-01', ho_days_in_future=None, spark=None):
+    """
+    holdout: whether to pull # of returns in holdout period. Useful
+        for evaluating model.
+    ho_days_in_future: int?; if 1, set the first day in holdout period
+        to be tomorrow. Negative numbers will set it to before today.
+    """
+    assert spark is not None, "Must pass SparkSession `spark`"
```

Comment:
The SparkSession should be an argument instead.

Reply:
If I'm reading this correctly, looks like you already did this in your PR by changing it from kwarg to 1st positional arg